feat(stats): Add data category to TrackOutcome

RyanSkonnord opened the pull request:
The TrackOutcome struct needs a DataCategory to indicate whether the
outcome is for an error, transaction, or attachment. Relay has the
notion of a `DataCategory` (and exports it in the `sentry-relay` Python
code) but doesn't apply it to outcomes.

Not every event emits an outcome from within the Relay component. The
immediate code changes are geared more toward transactions than
attachments. Transactions need little to no more than the category to
distinguish them, but attachments require more changes.

https://getsentry.atlassian.net/browse/ER-591

RyanSkonnord committed:
feat(stats): Add data category to TrackOutcome

mitsuhiko commented on relay-server/src/envelope.rs:
```diff
@@ -954,6 +968,27 @@ impl Envelope {
+            Some(data_categories[0])
+        } else {
+            if data_categories.len() > 1 {
+                relay_log::warn!("Items with conflicting data categories in the same envelope");
```
This would be very common for error + session for instance which come in the same envelope.

RyanSkonnord replied:
Thanks, I was hoping to investigate that in testing.

Downstream, we'll be interested mainly in the `Error`, `Transaction`, and `Attachment` categories. I considered filtering only to those values here, but that business logic doesn't belong in Relay. Having the outcome include the set of categories among all items in the envelope should be fine.

RyanSkonnord committed:
Include multiple data categories

RyanSkonnord committed:
Update integration test assertions

RyanSkonnord commented on tests/integration/test_outcome.py:
```diff
@@ -245,7 +246,10 @@ def test_outcome_source(relay, mini_sentry):
+    relay,
+    relay_with_processing,
+    outcomes_consumer,
+    num_intermediate_relays,
```
From "CI / Lint":
```
Run black --check py tests --exclude '\.eggs|sentry_relay/_lowlevel.*'
would reformat /home/runner/work/relay/relay/tests/integration/test_outcome.py
Oh no! 💥 💔 💥
```

~~I think there must be a version/config mismatch with Black, because my local Git hooks insisted I change it *to* this before committing.~~ (Resolved. I think I didn't have `relay/.venv/bin/black` activated correctly.)

markstory commented on tests/integration/test_outcome.py:
```diff
@@ -298,6 +302,7 @@ def test_outcome_forwarding(
         "reason": "project_id",
         "event_id": event_id,
         "remote_addr": "127.0.0.1",
+        "data_categories": ["error"],
```
If the event_id is a single event, would that event have more than one category?

jan-auer replied:
Good catch! There could be multiple outcomes, in fact, but for each outcome there should only ever be a single category. See one of the previous comments for an example.

jan-auer commented on relay-server/src/actors/outcome.rs:
```diff
@@ -81,6 +81,8 @@ pub struct TrackOutcome {
     /// The client ip address.
     pub remote_addr: Option<IpAddr>,
+    /// The data categories of events in the envelope.
+    pub data_categories: Vec<DataCategory>,
```
Every outcome should only ever have a single data category. If there were multiple categories, we should also emit multiple outcomes.

In the future, we will have to emit separate outcomes for each of the items in an envelope (that require outcomes). For example, you could be rate limiting an attachment, filtering the event, and accepting a session from a single envelope, and all with different reason codes.

For now, only the event item is relevant. Since there can only ever be a single event item (error, default, security, or transaction) in an envelope, it would be sufficient to use that and emit a single outcome.

jan-auer commented on relay-server/src/envelope.rs:
```diff
@@ -107,6 +109,18 @@ impl ItemType {
     }
+
+    /// Returns the data category type corresponding to this item type, if one exists.
+    fn as_data_category(&self) -> Option<DataCategory> {
```
This function already exists in `relay_server::utils` as `infer_event_category`; with the one difference that it doesn't return attachments. We use this for rate limiting purposes, currently. Would it be possible to use that function on the call site rather than adding it here?

jan-auer commented on relay-server/src/actors/outcome.rs:
```diff
@@ -304,6 +306,8 @@ pub struct TrackRawOutcome {
     #[serde(default, skip_serializing_if = "Option::is_none")]
     source: Option<String>,
+    /// The data categories of the events
+    pub data_categories: Vec<DataCategory>,
```
Please annotate this with `#[serde(default)]`. Usually, serde would require this new field and reject any past payload that we receive without. Since our PoP Relays are sending outcomes already without this, we need to gracefully accept them.

When going with a single DataCategory, you may have to specify an explicit default via `#[serde(default = "default_data_category")]` and then define a function:

```rust
fn default_data_category() -> DataCategory {
    DataCategory::Error
}
```

RyanSkonnord replied:
Will do. But (if it's no longer a `Vec`), would `Option<DataCategory>` be okay? That seems a natural fit with `infer_event_category`.

jan-auer reviewed (changes requested):
Outcomes in the data set are designed to carry only a single data category. For instance, in Sentry we create one outcome for the event and then another one per attachment. See one of my comments below for an example.

For now, I would propose to emit outcomes just for the event item. Every envelope can at most have a single event item (error, default, security, or transaction). That way, you can keep most of the code, and just have a single data category.

In a follow-up PR, you could emit separate outcomes for attachments, with the appropriate reason codes and data category. Doing outcomes for attachments is a whole different beast, as we'll have to hook into rate limiting and a few other places. Most likely, we'll have to change how we keep track of outcomes alltogether.

RyanSkonnord committed:
Put only one category in outcome, per infer_event_category

RyanSkonnord commented on relay-server/src/endpoints/common.rs:
```diff
@@ -526,6 +526,7 @@ where
                         outcome,
                         event_id: *event_id.borrow(),
                         remote_addr,
+                        event_category: None,
```
Just checking: my assumption here is that we failed to get any items from the envelope, therefore we can't possibly supply a category. Right?

RyanSkonnord commented on relay-server/src/envelope.rs:
```diff
@@ -954,6 +956,23 @@ impl Envelope {
+        if event_categories.len() == 1 {
+            Some(event_categories[0])
+        } else {
+            if event_categories.len() > 1 {
```
Let me know if there is a more graceful way to handle this than `collect()`'ing into a `Vec`. I'm still a newb with Rust's fluent iterators.

Unlike before, I'm more confident that we should never have to print this warning. Tell me if you think there's a better way to handle the condition -- perhaps returning `Some(event_categories[0])` would make more sense.

RyanSkonnord commented on relay-server/src/utils/rate_limits.rs:
```diff
@@ -85,7 +85,7 @@ pub fn parse_rate_limits(scoping: &Scoping, string: &str) -> RateLimits {
 /// - `Attachment`: If the attachment creates an event (e.g. for minidumps), the category is assumed
 ///   to be `Error`.
-fn infer_event_category(item: &Item) -> Option<DataCategory> {
+pub fn infer_event_category(item: &Item) -> Option<DataCategory> {
```
Rather than simply adding `pub`, would it be better to move this elsewhere? (With this change, it's arguably not strictly scoped to the concern of rate limits.)

jan-auer replied:
For simplicity, you can totally keep this here for this changeset. We'll likely have to touch this in a follow-up and create more utilities like this, at which point code organization will become more clear.

RyanSkonnord committed:
Simplify multiple event categories

RyanSkonnord committed:
Extract envelope's event category in handle_store_like_request

Checking for the presence of the event category should replace the
function's is_event parameter, which is removed.

RyanSkonnord commented on relay-server/src/endpoints/common.rs:
```diff
@@ -518,14 +519,16 @@ where
 
-            if is_event {
+            let event_category = *event_category.borrow();
+            if event_category.is_some() {
```
Requesting a Rust style check... is this the prettiest way to write this?

I considered
```rust
if let Some(event_category) = *event_category.borrow() {
```
but then line 531 needs to change to
```rust
event_category: Some(event_category),
```
which suggests that we don't want to unwrap the `Option` just to wrap it again.

Also, reusing the name `event_category` in a new `let` statement seems dubious by other languages' standards, but Rust seems to be cool with it and there's no better name for the value.

jan-auer replied:
> Also, reusing the name `event_category` in a new `let` statement seems dubious by other languages' standards, but Rust seems to be cool with it and there's no better name for the value.

Yeah. Thanks to strict typing and scoping rules we make use of this pattern in exactly such cases. Really all you do here is unbox a value, so I think it's totally fair 👍 

> I considered [...] but then line 531 needs to change to [...] `Some(event_category)` [...] which suggests that we don't want to unwrap the Option just to wrap it again.

That's precisely what you're doing, though. You're statically ensuring that you never pass `None` in. Using `Some(event_category)` below makes this even more explicit. It's not a bad thing to make it stick out that we really want to pass a `DataCategory` here. Both are equally fine to me, though, so your choice.

This gives me another idea. While we want to keep it an option on the raw outcome struct, have you considered making it non-optional on `TrackOutcome`? I believe at no place in the code we should be emitting `data_category: None` anymore if we're doing it right. This would be a neat way to enforce this.

jan-auer commented on relay-server/src/envelope.rs:
```diff
@@ -959,19 +958,9 @@ impl Envelope {
-        }
+        self.items().find_map(infer_event_category)
+        // There are some cases where multiple items may have different categories, but returning
+        // the first is good enough for now.
```
Agreed 👍  

The case we will have to look at is if someone is sending a `__sentry-event` msgpack attachment + a minidump. Depending on the order, we might get `::Default` or `::Error`. For stats, they are all equal, so it's not a big problem. 

`EnvelopeLimiter` already takes care of that, so it shouldn't be too hard to abstract that out in a follow-up.

jan-auer commented on relay-server/src/actors/events.rs:
```diff
@@ -1451,6 +1451,7 @@ impl Handler<HandleEnvelope> for EventManager {
             start_time,
             sampling_project,
         } = message;
+        let event_category = envelope.get_event_category();
```
If you look a few lines lower, there is the `is_event` flag with a comment on how this relates to outcomes. You can replace that with your new `event_category` option and keep the comment.

Particularly, you can replace the `if is_event` with the same `if let Some()` or `if event_category.is_some()` that you use in the common endpoint code.

jan-auer commented on relay-server/src/envelope.rs:
```diff
@@ -38,15 +38,16 @@ use std::io::{self, Write};
 use bytes::Bytes;
 use chrono::{DateTime, Utc};
 use failure::Fail;
+use relay_common::DataCategory;
```
_nit_: Please move this import to the next group where all the `relay_*` imports are.

jan-auer commented on relay-server/src/envelope.rs:
```diff
@@ -38,15 +38,16 @@ use std::io::{self, Write};
-use crate::constants::DEFAULT_EVENT_RETENTION;
 use crate::extractors::{PartialMeta, RequestMeta};
 use crate::utils::{ErrorBoundary, TraceContext};
+use crate::{constants::DEFAULT_EVENT_RETENTION, utils::infer_event_category};
```
_nit_: Looks like RustAnalyzer went rogue here. Could you manually re-group these imports? 

```rust
use crate::constants::DEFAULT_EVENT_RETENTION;
use crate::extractors::{PartialMeta, RequestMeta};
use crate::utils::{infer_event_category, ErrorBoundary, TraceContext};
```

jan-auer reviewed (commented):
This is great! I'm +1 on all your changes now, and everything links up correctly.

View final comments below, mostly for styling, and then we can merge this. 🚀

RyanSkonnord committed:
Tweaks from code review

jan-auer commented on relay-server/src/actors/outcome.rs:
```diff
@@ -304,6 +306,9 @@ pub struct TrackRawOutcome {
     source: Option<String>,
+    /// The data categories of the events
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub event_category: Option<DataCategory>,
```
Sorry, I'm just realizing that this needs to be called `category`. This directly maps to the schema on the Kafka topic, and both Sentry and the consumer already expect `category` for this.

RyanSkonnord committed:
Simplify "is event" check

RyanSkonnord committed:
Rename TrackOutcome category field for downstream consistency

RyanSkonnord committed:
Make outcome category non-optional

RyanSkonnord committed:
Make outcome optional in TrackRawOutcome again

Is this needed for compatibility with the outcomes endpoint?

RyanSkonnord commented on relay-server/src/actors/outcome.rs:
```diff
@@ -304,6 +306,9 @@ pub struct TrackRawOutcome {
     source: Option<String>,
+    /// The event's data category.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub category: Option<DataCategory>,
```
We're good with `TrackOutcome` having a required category, but does *this* one (on `TrackRawOutcome`) need to be optional? It seems that `TrackRawOutcome` is being used, in `relay-server/src/endpoints/outcomes.rs`, to deserialize outcomes from upstream Relays. If those Relays are self-hosted ones that aren't running the latest version, they might be sending JSON that lacks the `category` attribute, and we want to tolerate that, right? (Credit to @JoshFerge for the catch.)

I checked in b9de4819dc9ca2119ea4ed0700a2c762aa8b2085 on the hypothesis that this is correct. I'll revert it otherwise.

jan-auer committed:
test: Fix test assertion for attachment test

jan-auer committed:
Merge branch 'master' into outcome-data-category

* master:
  release: 21.2.0

jan-auer committed:
meta: Changelog

jan-auer reviewed (approved):
🍪

RyanSkonnord committed:
Add transaction event cases to test_outcome

RyanSkonnord committed:
Merge branch 'master' into outcome-data-category

jan-auer reviewed (approved):
Tests look good, let's merge this together.

jan-auer reviewed (changes requested):
Actually, we're missing a test case for processing mode.

jan-auer committed:
test: Test outcome categories in processing mode

jan-auer reviewed (approved):
Now we're good :)
